File: examples/runnable/09-retries-heartbeats/go/processing.go
// Package retries holds the retries + heartbeat lab. The Go equivalent of the
// Java retries lab. ChargeCard fails its first two attempts so the RetryPolicy is
// visible in history; ExportLargeReport heartbeats once per page so a Worker
// restart resumes mid-flight from the last recorded page instead of starting over.
//
// The Workflow and Activity definitions live here so both standalone commands
// can import them: ./worker (registers + polls) and ./starter (starts a run).
package retries

import (
	"context"
	"fmt"
	"log"
	"sync/atomic"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

const TaskQueue = "retries-heartbeats"

// FlakyActivities holds state shared across attempts on this Worker, so the
// charge counter survives retries (mirrors the Java AtomicInteger).
type FlakyActivities struct {
	chargeAttempts int32
}

// ChargeCard fails the first two attempts, then succeeds.
func (a *FlakyActivities) ChargeCard(ctx context.Context, orderID string) (string, error) {
	attempt := atomic.AddInt32(&a.chargeAttempts, 1)
	log.Printf("ChargeCard attempt %d for %s", attempt, orderID)
	if attempt < 3 {
		// A returned error is retried per the Workflow's RetryPolicy.
		return "", temporal.NewApplicationError(
			fmt.Sprintf("payment gateway timeout (attempt %d)", attempt), "GatewayTimeout")
	}
	return fmt.Sprintf("charged %s on attempt %d", orderID, attempt), nil
}

// ExportLargeReport heartbeats once per page so a restart resumes mid-flight.
func (a *FlakyActivities) ExportLargeReport(ctx context.Context, pages int) (string, error) {
	// On retry, resume from the last recorded heartbeat instead of page 0.
	startPage := 0
	if activity.HasHeartbeatDetails(ctx) {
		_ = activity.GetHeartbeatDetails(ctx, &startPage)
	}
	for page := startPage; page < pages; page++ {
		time.Sleep(300 * time.Millisecond)
		log.Printf("exported page %d/%d", page+1, pages)
		activity.RecordHeartbeat(ctx, page+1)
	}
	return fmt.Sprintf("exported %d pages", pages), nil
}

// ProcessingWorkflow charges the card (with retries) then exports the report
// (with a heartbeat timeout).
func ProcessingWorkflow(ctx workflow.Context, orderID string) (string, error) {
	// A deliberate retry policy: 5 attempts, 1s initial backoff doubling each time.
	chargeCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    time.Second,
			BackoffCoefficient: 2.0,
			MaximumAttempts:    5,
		},
	})

	var a *FlakyActivities
	var charge string
	if err := workflow.ExecuteActivity(chargeCtx, a.ChargeCard, orderID).Get(ctx, &charge); err != nil {
		return "", err
	}

	// A long Activity: the heartbeat timeout detects a dead Worker between pages.
	reportCtx := workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Minute,
		HeartbeatTimeout:    5 * time.Second,
	})
	var exported string
	if err := workflow.ExecuteActivity(reportCtx, a.ExportLargeReport, 5).Get(ctx, &exported); err != nil {
		return "", err
	}

	return charge + " | " + exported, nil
}

File: examples/runnable/09-retries-heartbeats/python/processing.py
"""Retry + heartbeat deep-dive. The Python equivalent of the Java retries lab.

chargeCard fails its first two attempts so the RetryPolicy is visible in history;
exportLargeReport heartbeats once per page so a Worker restart resumes mid-flight
from the last recorded page instead of starting over.
"""

import asyncio
from datetime import timedelta

from temporalio import activity, workflow
from temporalio.common import RetryPolicy
from temporalio.exceptions import ApplicationError

TASK_QUEUE = "retries-heartbeats"


class FlakyActivities:
    def __init__(self) -> None:
        # One instance is shared across attempts on this Worker, so the counter
        # survives retries (mirrors the Java AtomicInteger).
        self._charge_attempts = 0

    @activity.defn
    async def charge_card(self, order_id: str) -> str:
        self._charge_attempts += 1
        attempt = self._charge_attempts
        print(f"charge_card attempt {attempt} for {order_id}")
        if attempt < 3:
            # A raised error is retried per the Workflow's retry_policy.
            raise ApplicationError(
                f"payment gateway timeout (attempt {attempt})", type="GatewayTimeout"
            )
        return f"charged {order_id} on attempt {attempt}"

    @activity.defn
    async def export_large_report(self, pages: int) -> str:
        # On retry, resume from the last recorded heartbeat instead of page 0.
        start_page = 0
        if activity.info().heartbeat_details:
            start_page = activity.info().heartbeat_details[0]
        for page in range(start_page, pages):
            await asyncio.sleep(0.3)
            print(f"exported page {page + 1}/{pages}")
            activity.heartbeat(page + 1)
        return f"exported {pages} pages"


@workflow.defn
class ProcessingWorkflow:
    @workflow.run
    async def process(self, order_id: str) -> str:
        # A deliberate retry policy: 5 attempts, 1s initial backoff doubling each time.
        charge = await workflow.execute_activity_method(
            FlakyActivities.charge_card,
            order_id,
            start_to_close_timeout=timedelta(seconds=10),
            retry_policy=RetryPolicy(
                initial_interval=timedelta(seconds=1),
                backoff_coefficient=2.0,
                maximum_attempts=5,
            ),
        )

        # A long Activity: the heartbeat timeout detects a dead Worker between pages.
        exported = await workflow.execute_activity_method(
            FlakyActivities.export_large_report,
            5,
            start_to_close_timeout=timedelta(minutes=5),
            heartbeat_timeout=timedelta(seconds=5),
        )

        return f"{charge} | {exported}"
